### PR TITLE
Switch to use server side schema description for KeyspaceMetadata.ToCQL

### PR DESCRIPTION
**Pull request by dkropachev**
No need to keep client-side logic which needs to be updated every time there is changes on server-side.

Closes https://github.com/scylladb/gocql/issues/195

### CODE REVIEW

**Comment by sylwiaszunejko** [comment]
Does it work against all supported versions of Scylla? Fallback to the old logic is not necessary?

**Comment by dkropachev** [comment]
> Does it work against all supported versions of Scylla? Fallback to the old logic is not necessary?

Tested on both old scylla and old cassandra, in both cases `errFrame` code ends up being a `ErrCodeSyntax`.
Added logic to backup to construct statement on client side in such case.

**Review by sylwiaszunejko — DISMISSED** [review]

**Comment by sylwiaszunejko** [comment]
I think it LGTM, @wprzytula maybe you could give a second opinion?

File: recreate_test.go [hunk 1]
```diff
@@ -97,13 +101,20 @@ func TestRecreateSchema(t *testing.T) {
 				t.Fatal("recreate schema", err)
 			}
 
+			dump = trimSchema(dump)
+			if *updateGolden {
+				if err := ioutil.WriteFile(test.Golden, []byte(dump), 0644); err != nil {
+					t.Fatal(err)
+				}
+			}
+
```
Comment (wprzytula) [inline recreate_test.go:110]: Why are these changes here???

**Review by wprzytula — CHANGES_REQUESTED** [review]

File: recreate_test.go (hunk 1 above)
Comment (dkropachev) [reply recreate_test.go:110]: To make it possible to update golden files and to make test properly match new ToCQL() return

File: testdata/recreate/aggregates_golden.cql [hunk 2]
```diff
@@ -1,33 +1,25 @@
...
-  return { state[1]+1, state[2]+val }
-  $$;
-
-CREATE FUNCTION gocqlx_aggregates.avgfinal (state
-    tuple<int, double>)
-    CALLED ON NULL INPUT
-    RETURNS double
-    LANGUAGE lua
-    AS $$
+CREATE KEYSPACE gocqlx_aggregates WITH replication = {'class': 'org.apache.cassandra.locator.SimpleStrategy', 'replication_factor': '2'} AND durable_writes = true;
```
[the remark this answers is not in the document]
Comment (dkropachev) [reply testdata/recreate/aggregates_golden.cql:1]: It is expected result for the given test.

File: testdata/recreate/udt_golden.cql [hunk 3]
```diff
@@ -1,16 +1,11 @@
-CREATE KEYSPACE gocqlx_udt WITH replication = {
-    'class': 'SimpleStrategy',
-    'replication_factor': '2'
-};
-
+CREATE KEYSPACE gocqlx_udt WITH replication = {'class': 'org.apache.cassandra.locator.SimpleStrategy', 'replication_factor': '2'} AND durable_writes = true;
```
[the remark this answers is not in the document]
Comment (dkropachev) [reply testdata/recreate/udt_golden.cql:1]: This is just expected result of .ToCQL(), schema is in other cql files, `tcs[id].Input`.
For sake of not polluting this PR, let's have a separate issue on switching from `SimpleStrategy` to `NetworkTopologyStrategy`

File: testdata/recreate/table_golden.cql [hunk 4]
```diff
@@ -11,36 +7,39 @@ CREATE TABLE gocqlx_table.loads (
...
     AND default_time_to_live = 0
     AND gc_grace_seconds = 864000
     AND max_index_interval = 2048
     AND memtable_flush_period_in_ms = 0
     AND min_index_interval = 128
-    AND speculative_retry = '99.0PERCENTILE';
-
+    AND speculative_retry = '99.0PERCENTILE'
+    AND paxos_grace_seconds = 864000
+    AND tombstone_gc = {'mode': 'timeout', 'propagation_delay_in_seconds': '3600'};
```
[the remark this answers is not in the document]
Comment (dkropachev) [reply testdata/recreate/table_golden.cql:42]: This is just expected result of .ToCQL(), schema is in other cql files, `tcs[id].Input`.
For sake of not polluting this PR, let's have a separate issue on switching from `SimpleStrategy` to `NetworkTopologyStrategy`
Since now we read it from server side result has changed.

File: testdata/recreate/table_golden.cql [hunk 5]
```diff
@@ -1,8 +1,4 @@
-CREATE KEYSPACE gocqlx_table WITH replication = {
-    'class': 'SimpleStrategy',
-    'replication_factor': '2'
-};
-
+CREATE KEYSPACE gocqlx_table WITH replication = {'class': 'org.apache.cassandra.locator.SimpleStrategy', 'replication_factor': '2'} AND durable_writes = true;
```
[the remark this answers is not in the document]
Comment (dkropachev) [reply testdata/recreate/table_golden.cql:1]: This is just expected result of .ToCQL(), schema is in other cql files, `tcs[id].Input`.
For sake of not polluting this PR, let's have a separate issue on switching from `SimpleStrategy` to `NetworkTopologyStrategy`

File: metadata_scylla.go [hunk 6]
```diff
@@ -744,6 +753,36 @@ func getIndexMetadata(session *Session, keyspaceName string) ([]IndexMetadata, e
 	return indexes, nil
 }
 
+// get create statements for the keyspace
+func getCreateStatements(session *Session, keyspaceName string) ([]string, error) {
+	if !session.useSystemSchema {
+		return nil, nil
+	}
```
[the remark this answers is not in the document]
Comment (dkropachev) [reply metadata_scylla.go:760]: there is fallback logic in `ToCQL()`

File: metadata_scylla.go [hunk 7]
```diff
@@ -744,6 +753,36 @@ func getIndexMetadata(session *Session, keyspaceName string) ([]IndexMetadata, e
...
+	var createStatements []string
+	for _, row := range rows {
+		stmt := row["create_statement"]
+		if stmt == nil {
+			continue
+		}
+		strStmt, ok := stmt.(string)
+		if !ok {
+			continue
+		}
```
[the remark this answers is not in the document]
Comment (dkropachev) [reply metadata_scylla.go]: Reimplemented, now there is no guessing.

File: testdata/recreate/udt_golden.cql (hunk 3 above)
[the remark this answers is not in the document]
Comment (wprzytula) [reply testdata/recreate/udt_golden.cql:1]: Will you open the issue?

File: testdata/recreate/table_golden.cql (hunk 4 above)
[the remark this answers is not in the document]
Comment (wprzytula) [reply testdata/recreate/table_golden.cql:42]: Are we 100% sure that in tests we would always use the server-side schema description? (i.e. under no circumstances will fallback to client-side description fire up)

File: metadata_scylla.go [hunk 8]
```diff
@@ -744,6 +753,36 @@ func getIndexMetadata(session *Session, keyspaceName string) ([]IndexMetadata, e
 	return indexes, nil
 }
 
+// get create statements for the keyspace
+func getCreateStatements(session *Session, keyspaceName string) ([]string, error) {
+	if !session.useSystemSchema {
+		return nil, nil
+	}
+	rows, err := session.control.query(fmt.Sprintf(`DESCRIBE KEYSPACE %s WITH INTERNALS`, keyspaceName)).SliceMap()
```
[the remark this answers is not in the document]
Comment (wprzytula) [reply metadata_scylla.go]: So where did you get this syntax from?

File: metadata_scylla.go [hunk 9]
```diff
@@ -744,6 +753,36 @@ func getIndexMetadata(session *Session, keyspaceName string) ([]IndexMetadata, e
...
+	for _, row := range rows {
+		stmt := row["create_statement"]
+		if stmt == nil {
+			continue
+		}
+		strStmt, ok := stmt.(string)
+		if !ok {
+			continue
+		}
+		createStatements = append(createStatements, strStmt)
```
[the remark this answers is not in the document]
Comment (wprzytula) [reply metadata_scylla.go]: IIUC, `append` reallocates the underlying array in place. Or not? Does it always allocate a separate array, never extends the previous allocation?

[the remark this answers is not in the document]
Comment (dkropachev) [reply metadata_scylla.go]: It does reallocate underlying array, but only in case if it can fit whatever you apending.

File: metadata_scylla.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (dkropachev) [reply metadata_scylla.go]: https://github.com/scylladb/gocql/issues/195

File: testdata/recreate/udt_golden.cql (hunk 3 above)
[the remark this answers is not in the document]
Comment (dkropachev) [reply testdata/recreate/udt_golden.cql:1]: https://github.com/scylladb/gocql/issues/217

**Comment by sylwiaszunejko** [comment]
@wprzytula ping

File: testdata/recreate/table_golden.cql (hunk 4 above)
[the remark this answers is not in the document]
Comment (dkropachev) [reply testdata/recreate/table_golden.cql:42]: I will take care of it

**Comment by dkropachev** [comment]
> @wprzytula ping

I need to address one problem he pointed out

File: testdata/recreate/table_golden.cql (hunk 4 above)
[the remark this answers is not in the document]
Comment (dkropachev) [reply testdata/recreate/table_golden.cql]: I have added logic to the test to handle both cases.

**Comment by wprzytula** [comment]
> > @wprzytula ping
> 
> I need to address one problem he pointed out

Has this been resolved?

**Review by wprzytula — APPROVED** [review]
